contracts/Garrison.sol

```diff
@@ -0,0 +1,134 @@
+pragma solidity ^0.6.2;
+
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721ReceiverUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/utils/EnumerableSet.sol";
+
+import "./characters.sol";
+
+contract Garrison is Initializable, IERC721ReceiverUpgradeable, AccessControlUpgradeable {
+    using EnumerableSet for EnumerableSet.UintSet;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    bytes32 public constant GAME_ADMIN = keccak256("GAME_ADMIN");
+
+    // STATE
+    Characters characters;
+
+    EnumerableSet.AddressSet private supportedTokenTypes;
+
+    mapping(address => EnumerableSet.UintSet) userGarrison;
+    mapping(uint256 => address) characterOwner;
+    EnumerableSet.UintSet private allCharactersInGarrison;
+
+    event CharacterReceived(uint256 indexed character, address indexed minter);
+
+    function initialize(Characters _characters)
+        public
+        initializer
+    {
+        __AccessControl_init();
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        characters = _characters;
+    }
+
+    // MODIFIERS
+    modifier restricted() {
+        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender) || hasRole(GAME_ADMIN, msg.sender), "Not admin");
+        _;
+    }
+
+    modifier isCharacterOwner(uint256 id) {
+        require(characterOwner[id] == msg.sender);
+        _;
+    }
+
+    modifier isInGarrison(uint256 id) {
+        require(allCharactersInGarrison.contains(id));
+        _;
+    }
+
+    // VIEWS
+    function getUserCharacters() public view  returns (uint256[] memory tokens) {
+        uint256 amount = balanceOf(msg.sender);
+        tokens = new uint256[](amount);
+
+        EnumerableSet.UintSet storage storedTokens = userGarrison[msg.sender];
+
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 id = storedTokens.at(i);
+                tokens[i] = id;
+        }
+    }
+
+    function balanceOf(address user) public view returns(uint256) {
+        return userGarrison[user].length();
+    }
+
+    // MUTATIVE
+    function sendToGarrison(uint256 id) public {
```

Comment:
Shouldn't there be some restrictions? Like checks, if the id provided is user's character? This method is public

Reply:
Not needed, it will just fail on safeTransferFrom if sender is not the owner of the nft. We could add a require statement just to throw an explicit error about what happened, but there's no way that happens when using our frontend for this call.